File: src/io/mdf-reader.js
// ASAM MDF v4 (.mf4/.mdf) reader — the dominant automotive measurement format
// (CANape, INCA/ETAS, Vector bus logging, PEAK). We read the CAN bus-logging
// layout: channel groups whose channels follow the standard "CAN_DataFrame"
// (and "CAN_ErrorFrame") naming, decoding each record into a FrameStore.
//
// Scope: sorted and unsorted data groups; uncompressed (##DT/##DV), compressed
// (##DZ, deflate, incl. transposed), and block lists (##DL/##HL). Generic
// signal decoding (non-bus MDF), array/VLSD channels and MDF3 are out of scope
// — those would surface as "no CAN frames found".

import { BinaryReader, inflate } from '../util/binary.js';
import { FrameStore, FrameFlags } from '../core/frame-store.js';
import { registerFormat } from './format-registry.js';

const TEXT = new TextDecoder();

/** Common v4 block header: id "##XY", length, link section. */
function readBlockHeader(bytes, offset) {
  const id = String.fromCharCode(bytes[offset], bytes[offset + 1], bytes[offset + 2], bytes[offset + 3]);
  const dv = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  const length = Number(dv.getBigUint64(offset + 8, true));
  const linkCount = Number(dv.getBigUint64(offset + 16, true));
  const links = new Array(linkCount);
  for (let i = 0; i < linkCount; i++) links[i] = Number(dv.getBigUint64(offset + 24 + i * 8, true));
  const dataStart = offset + 24 + linkCount * 8;
  return { id, length, linkCount, links, dataStart, end: offset + length };
}

function blockReaderAt(bytes, offset) {
  const h = readBlockHeader(bytes, offset);
  return { h, r: new BinaryReader(bytes, h.dataStart) };
}

/** UTF-8 text of a ##TX / ##MD block, or '' for a null link. */
function readText(bytes, offset) {
  if (!offset) return '';
  const h = readBlockHeader(bytes, offset);
  if (h.id !== '##TX' && h.id !== '##MD') return '';
  const raw = bytes.subarray(h.dataStart, h.end);
  return TEXT.decode(raw).replace(/\0.*$/s, '').trim();
}

/** Resolve a data-block reference (##DT/##DV/##DZ/##DL/##HL) to a byte array. */
async function readDataBlock(bytes, offset) {
  if (!offset) return new Uint8Array(0);
  const h = readBlockHeader(bytes, offset);
  switch (h.id) {
    case '##DT':
    case '##DV':
    case '##RD':
      return bytes.subarray(h.dataStart, h.end);
    case '##DZ':
      return inflateDz(bytes, h);
    case '##DL': {
      // Data list: concatenate all referenced blocks, following dl_dl_next.
      const parts = [];
      let cur = offset;
      while (cur) {
        const hh = readBlockHeader(bytes, cur);
        const r = new BinaryReader(bytes, hh.dataStart);
        r.u8(); r.skip(3); // dl_flags + reserved
        const count = r.u32();
        for (let i = 0; i < count; i++) parts.push(await readDataBlock(bytes, hh.links[1 + i]));
        cur = hh.links[0]; // dl_dl_next
      }
      return concat(parts);
    }
    case '##HL': {
      // Header list -> points at the first ##DL.
      return readDataBlock(bytes, h.links[0]);
    }
    default:
      return new Uint8Array(0);
  }
}

async function inflateDz(bytes, h) {
  const r = new BinaryReader(bytes, h.dataStart);
  r.skip(2); // dz_org_block_type (2 chars)
  const zipType = r.u8();
  r.u8(); // reserved
  const zipParam = r.u32();
  const orgLen = Number(r.u64());
  const dataLen = Number(r.u64());
  const comp = bytes.subarray(r.pos, r.pos + dataLen);
  let out = await inflate(comp);
  if (out.length !== orgLen) out = out.subarray(0, orgLen);
  if (zipType === 1) out = untranspose(out, zipParam, orgLen);
  return out;
}

/** Undo MDF column-wise transposition (zip_type 1). */
function untranspose(data, cols, orgLen) {
  if (!cols) return data;
  const rows = Math.floor(orgLen / cols);
  const out = new Uint8Array(orgLen);
  for (let c = 0; c < cols; c++) {
    for (let rrow = 0; rrow < rows; rrow++) {
      out[rrow * cols + c] = data[c * rows + rrow];
    }
  }
  const tail = rows * cols;
  if (tail < orgLen) out.set(data.subarray(tail), tail);
  return out;
}

function concat(parts) {
  let total = 0;
  for (const p of parts) total += p.length;
  const out = new Uint8Array(total);
  let o = 0;
  for (const p of parts) { out.set(p, o); o += p.length; }
  return out;
}

/** Parse a ##CN channel block at offset. */
function readChannel(bytes, offset) {
  const h = readBlockHeader(bytes, offset);
  const r = new BinaryReader(bytes, h.dataStart);
  const cn = {
    next: h.links[0],
    composition: h.links[1],
    txName: h.links[2],
    ccConversion: h.links[4],
    type: r.u8(),
    syncType: r.u8(),
    dataType: r.u8(),
    bitOffset: r.u8(),
    byteOffset: r.u32(),
    bitCount: r.u32(),
  };
  cn.name = readText(bytes, cn.txName);
  return cn;
}

/** Read a linear/identity conversion; returns {a,b} for phys = a + b*raw. */
function readConversion(bytes, offset) {
  if (!offset) return null;
  const h = readBlockHeader(bytes, offset);
  if (h.id !== '##CC') return null;
  const r = new BinaryReader(bytes, h.dataStart);
  const ccType = r.u8();
  r.u8(); r.u16(); // precision, flags
  r.u16(); // ref_count
  const valCount = r.u16();
  r.f64(); r.f64(); // phy range min/max
  const vals = [];
  for (let i = 0; i < valCount; i++) vals.push(r.f64());
  if (ccType === 0) return { a: 0, b: 1 }; // identity
  if (ccType === 1) return { a: vals[0] ?? 0, b: vals[1] ?? 1 }; // linear
  return null; // other conversions not needed for bus timestamps
}

/** Collect all channels of a CG (following cn_cn_next and cn_composition). */
function collectChannels(bytes, firstCn) {
  const list = [];
  const walk = (offset) => {
    let cur = offset;
    while (cur) {
      const cn = readChannel(bytes, cur);
      list.push(cn);
      if (cn.composition) {
        const ch = readBlockHeader(bytes, cn.composition);
        if (ch.id === '##CN') walk(cn.composition); // nested channels
        // ##CA (arrays) are not expanded — DataBytes is read as a byte array.
      }
      cur = cn.next;
    }
  };
  walk(firstCn);
  return list;
}

// Numeric field read from a record (byte-aligned little/big-endian ints & floats).
function readField(rec, base, cn) {
  const off = base + cn.byteOffset;
  const dv = new DataView(rec.buffer, rec.byteOffset, rec.byteLength);
  const bits = cn.bitCount;
  const dt = cn.dataType;
  // IEEE float
  if (dt === 4) return bits === 32 ? dv.getFloat32(off, true) : dv.getFloat64(off, true);
  if (dt === 5) return bits === 32 ? dv.getFloat32(off, false) : dv.getFloat64(off, false);
  const le = dt === 0 || dt === 2;
  const signed = dt === 2 || dt === 3;
  const nbytes = Math.ceil(bits / 8);
  let v = 0n;
  for (let i = 0; i < nbytes; i++) {
    const b = BigInt(rec[off + (le ? i : nbytes - 1 - i)]);
    v |= b << BigInt(8 * i);
  }
  // mask to bit count and shift by intra-byte bit offset if any
  if (cn.bitOffset) v >>= BigInt(cn.bitOffset);
  if (bits % 8 !== 0) v &= (1n << BigInt(bits)) - 1n;
  if (signed && (v & (1n << BigInt(bits - 1)))) v -= 1n << BigInt(bits);
  return Number(v);
}

export async function parseMdf(buffer) {
  const bytes = buffer instanceof Uint8Array ? buffer : new Uint8Array(buffer);
  const idMagic = String.fromCharCode(bytes[0], bytes[1], bytes[2]);
  if (idMagic !== 'MDF' && idMagic !== 'UnF') throw new Error('Not an MDF file');

  const store = new FrameStore();
  // HD block sits right after the 64-byte ID block.
  const hd = blockReaderAt(bytes, 64);
  if (hd.h.id !== '##HD') throw new Error('MDF: missing HD block');
  const startNs = Number(hd.r.u64());
  if (startNs > 0) store.t0Epoch = startNs / 1e9;

  let dgOff = hd.h.links[0]; // hd_dg_first
  let total = 0;
  while (dgOff) {
    const dg = blockReaderAt(bytes, dgOff);
    const dgNext = dg.h.links[0];
    const cgFirst = dg.h.links[1];
    const dgData = dg.h.links[2];
    const recIdSize = dg.r.u8();

    // Gather channel groups in this data group.
    const groups = [];
    let cgOff = cgFirst;
    while (cgOff) {
      const cg = blockReaderAt(bytes, cgOff);
      const cgNext = cg.h.links[0];
      const cnFirst = cg.h.links[1];
      const acqName = readText(bytes, cg.h.links[2]);
      const recordId = Number(cg.r.u64());
      const cycleCount = Number(cg.r.u64());
      const cgFlags = cg.r.u16();
      cg.r.u16(); // path separator
      cg.r.u32(); // reserved
      const dataBytes = cg.r.u32();
      const invalBytes = cg.r.u32();
      const isVlsd = (cgFlags & 0x01) !== 0;
      const channels = isVlsd ? [] : collectChannels(bytes, cnFirst);
      groups.push({ recordId, cycleCount, dataBytes, invalBytes, recSize: dataBytes + invalBytes, channels, acqName, isVlsd });
      cgOff = cgNext;
    }

    const data = await readDataBlock(bytes, dgData);
    total += decodeDataGroup(bytes, store, data, groups, recIdSize);
    dgOff = dgNext;
  }

  if (total === 0) {
    throw new Error(
      'MDF: no CAN frames found. This reader supports CAN bus-logging files ' +
      '(CAN_DataFrame channels); generic signal MDF is not supported.',
    );
  }
  return store;
}

function classifyGroup(bytes, g) {
  const find = (suffix) => g.channels.find((c) => {
    const leaf = c.name.toLowerCase().split('.').pop();
    return leaf === suffix;
  });
  const master = g.channels.find((c) => c.type === 2 || c.type === 3) || find('t');
  const id = find('id');
  const dataBytes = g.channels.find((c) => /databytes$/i.test(c.name.split('.').pop()));
  const nameHasCan = g.acqName.toLowerCase().includes('can') ||
    g.channels.some((c) => c.name.toLowerCase().includes('can_'));
  const isError = /errorframe/i.test(g.acqName) ||
    g.channels.some((c) => /errorframe/i.test(c.name));
  if (!master || (!id && !isError)) return null;
  if (!nameHasCan && !id) return null;
  return {
    master, id, dataBytes,
    ide: find('ide'),
    dlc: find('dlc'),
    dataLength: find('datalength'),
    dir: find('dir'),
    fd: find('edl') || find('fdf'),
    brs: find('brs'),
    esi: find('esi'),
    bus: find('buschannel'),
    isError,
    masterConv: readConversion(bytes, master.ccConversion),
  };
}

function decodeDataGroup(bytes, store, data, groups, recIdSize) {
  // Map record id -> group for unsorted data groups.
  const byId = new Map();
  for (const g of groups) byId.set(g.recordId, g);
  // Precompute CAN classification per group.
  for (const g of groups) g._can = g.isVlsd ? null : classifyGroup(bytes, g);

  let added = 0;
  let pos = 0;
  const n = data.length;

  const readRecId = () => {
    let id = 0;
    for (let i = 0; i < recIdSize; i++) id += data[pos + i] * 2 ** (8 * i);
    pos += recIdSize;
    return id;
  };

  if (recIdSize === 0) {
    // Sorted: exactly one channel group; records are back to back.
    const g = groups[0];
    if (!g) return 0;
    for (let k = 0; k < g.cycleCount && pos + g.recSize <= n; k++) {
      added += emitRecord(store, data, pos, g);
      pos += g.recSize;
    }
    return added;
  }

  // Unsorted: each record is prefixed by its group's record id.
  while (pos + recIdSize <= n) {
    const rid = readRecId();
    const g = byId.get(rid);
    if (!g) break; // desync — stop rather than misread
    if (pos + g.recSize > n) break;
    added += emitRecord(store, data, pos, g);
    pos += g.recSize;
  }
  return added;
}

function emitRecord(store, data, base, g) {
  const can = g._can;
  if (!can) return 0;
  const rec = data;
  const tRaw = readField(rec, base, can.master);
  const t = can.masterConv ? can.masterConv.a + can.masterConv.b * tRaw : tRaw;

  if (can.isError && !can.id) {
    store.add(t, 0, false, can.bus ? readField(rec, base, can.bus) : 0, FrameFlags.ERR, null);
    return 1;
  }

  let idVal = readField(rec, base, can.id) >>> 0;
  let ext;
  if (can.ide) ext = readField(rec, base, can.ide) !== 0;
  else { ext = (idVal & 0x80000000) !== 0; idVal &= 0x1fffffff; }
  idVal &= 0x1fffffff;

  const ch = can.bus ? readField(rec, base, can.bus) : 0;
  let flags = 0;
  if (can.isError) flags |= FrameFlags.ERR;
  if (can.fd && readField(rec, base, can.fd)) flags |= FrameFlags.FD;
  if (can.brs && readField(rec, base, can.brs)) flags |= FrameFlags.BRS;
  if (can.esi && readField(rec, base, can.esi)) flags |= FrameFlags.ESI;
  if (can.dir && readField(rec, base, can.dir)) flags |= FrameFlags.TX;

  // Actual payload length: DataLength if present, else DLC, else DataBytes width.
  let len;
  if (can.dataLength) len = readField(rec, base, can.dataLength);
  else if (can.dlc) len = readField(rec, base, can.dlc);
  else len = can.dataBytes ? can.dataBytes.bitCount / 8 : 0;
  len = Math.max(0, Math.min(64, len | 0));

  let payload = null;
  if (can.dataBytes) {
    const off = base + can.dataBytes.byteOffset;
    const width = can.dataBytes.bitCount / 8;
    payload = data.subarray(off, off + Math.min(len, width));
  }
  store.add(t, idVal, ext, ch, flags, payload);
  return 1;
}

registerFormat({
  name: 'ASAM MDF v4 (MF4)',
  extensions: ['mf4', 'mdf'],
  sniff(head) {
    // "MDF     " (finalized) or "UnFinMF " (unfinalized) at offset 0.
    return head[0] === 0x4d && head[1] === 0x44 && head[2] === 0x46;
  },
  async read(file) {
    const buf = new Uint8Array(await file.arrayBuffer());
    return parseMdf(buf);
  },
});

File: python/canviewer/io/mdf_reader.py
"""ASAM MDF v4 (.mf4) reader — CAN bus-logging layout. Port of the JS mdf-reader.

Reads CAN_DataFrame / CAN_ErrorFrame channel groups: sorted and unsorted data
groups; uncompressed (##DT/##DV), compressed (##DZ, deflate incl. transposed),
and block-list (##DL/##HL) data. Generic signal MDF, array/VLSD channels and
MDF3 are out of scope.
"""
from __future__ import annotations

import struct
import zlib

from ..core.frame_store import FrameStore, FD, BRS, ESI, ERR, TX, dlc_to_length
from .registry import Format, register_format


def _hdr(b, off):
    ident = b[off:off + 4].decode("latin-1")
    length = struct.unpack_from("<Q", b, off + 8)[0]
    link_count = struct.unpack_from("<Q", b, off + 16)[0]
    links = list(struct.unpack_from("<%dQ" % link_count, b, off + 24)) if link_count else []
    data_start = off + 24 + link_count * 8
    return {"id": ident, "length": length, "links": links,
            "data": data_start, "end": off + length}


def _text(b, off):
    if not off:
        return ""
    h = _hdr(b, off)
    if h["id"] not in ("##TX", "##MD"):
        return ""
    return b[h["data"]:h["end"]].split(b"\x00", 1)[0].decode("utf-8", "replace").strip()


def _read_data_block(b, off):
    if not off:
        return b""
    h = _hdr(b, off)
    ident = h["id"]
    if ident in ("##DT", "##DV", "##RD"):
        return b[h["data"]:h["end"]]
    if ident == "##DZ":
        return _inflate_dz(b, h)
    if ident == "##DL":
        parts = []
        cur = off
        while cur:
            hh = _hdr(b, cur)
            count = struct.unpack_from("<I", b, hh["data"] + 4)[0]
            for i in range(count):
                parts.append(_read_data_block(b, hh["links"][1 + i]))
            cur = hh["links"][0]
        return b"".join(parts)
    if ident == "##HL":
        return _read_data_block(b, h["links"][0])
    return b""


def _inflate_dz(b, h):
    o = h["data"]
    zip_type = b[o + 2]
    zip_param = struct.unpack_from("<I", b, o + 4)[0]
    org_len = struct.unpack_from("<Q", b, o + 8)[0]
    data_len = struct.unpack_from("<Q", b, o + 16)[0]
    comp = b[o + 24:o + 24 + data_len]
    out = zlib.decompress(comp)
    if len(out) > org_len:
        out = out[:org_len]
    if zip_type == 1:
        out = _untranspose(out, zip_param, org_len)
    return out


def _untranspose(data, cols, org_len):
    if not cols:
        return data
    rows = org_len // cols
    out = bytearray(org_len)
    for c in range(cols):
        for r in range(rows):
            out[r * cols + c] = data[c * rows + r]
    tail = rows * cols
    if tail < org_len:
        out[tail:] = data[tail:org_len]
    return bytes(out)


def _read_channel(b, off):
    h = _hdr(b, off)
    o = h["data"]
    cn = {
        "next": h["links"][0],
        "composition": h["links"][1],
        "cc": h["links"][4] if len(h["links"]) > 4 else 0,
        "type": b[o],
        "data_type": b[o + 2],
        "bit_offset": b[o + 3],
        "byte_offset": struct.unpack_from("<I", b, o + 4)[0],
        "bit_count": struct.unpack_from("<I", b, o + 8)[0],
    }
    cn["name"] = _text(b, h["links"][2])
    return cn


def _read_conversion(b, off):
    if not off:
        return None
    h = _hdr(b, off)
    if h["id"] != "##CC":
        return None
    o = h["data"]
    cc_type = b[o]
    val_count = struct.unpack_from("<H", b, o + 6)[0]
    vals = list(struct.unpack_from("<%dd" % val_count, b, o + 24)) if val_count else []
    if cc_type == 0:
        return (0.0, 1.0)
    if cc_type == 1:
        return (vals[0] if len(vals) > 0 else 0.0, vals[1] if len(vals) > 1 else 1.0)
    return None


def _collect_channels(b, first):
    out = []

    def walk(off):
        cur = off
        while cur:
            cn = _read_channel(b, cur)
            out.append(cn)
            if cn["composition"]:
                ch = _hdr(b, cn["composition"])
                if ch["id"] == "##CN":
                    walk(cn["composition"])
            cur = cn["next"]

    walk(first)
    return out


def _read_field(rec, base, cn):
    off = base + cn["byte_offset"]
    bits = cn["bit_count"]
    dt = cn["data_type"]
    if dt == 4:
        return struct.unpack_from("<f" if bits == 32 else "<d", rec, off)[0]
    if dt == 5:
        return struct.unpack_from(">f" if bits == 32 else ">d", rec, off)[0]
    le = dt in (0, 2)
    signed = dt in (2, 3)
    nbytes = (bits + 7) // 8
    chunk = rec[off:off + nbytes]
    if len(chunk) < nbytes:
        chunk = chunk + b"\x00" * (nbytes - len(chunk))
    v = int.from_bytes(chunk, "little" if le else "big")
    if cn["bit_offset"]:
        v >>= cn["bit_offset"]
    if bits % 8:
        v &= (1 << bits) - 1
    if signed and (v & (1 << (bits - 1))):
        v -= 1 << bits
    return v


def parse_mdf(b: bytes) -> FrameStore:
    if b[:3] not in (b"MDF", b"UnF"):
        raise ValueError("Not an MDF file")
    store = FrameStore()
    hd = _hdr(b, 64)
    if hd["id"] != "##HD":
        raise ValueError("MDF: missing HD block")
    start_ns = struct.unpack_from("<Q", b, hd["data"])[0]
    if start_ns > 0:
        store.t0_epoch = start_ns / 1e9

    total = 0
    dg_off = hd["links"][0]
    while dg_off:
        dg = _hdr(b, dg_off)
        dg_next, cg_first, dg_data = dg["links"][0], dg["links"][1], dg["links"][2]
        rec_id_size = b[dg["data"]]
        groups = []
        cg_off = cg_first
        while cg_off:
            cg = _hdr(b, cg_off)
            o = cg["data"]
            record_id = struct.unpack_from("<Q", b, o)[0]
            cycle_count = struct.unpack_from("<Q", b, o + 8)[0]
            cg_flags = struct.unpack_from("<H", b, o + 16)[0]
            data_bytes = struct.unpack_from("<I", b, o + 24)[0]
            inval_bytes = struct.unpack_from("<I", b, o + 28)[0]
            is_vlsd = bool(cg_flags & 0x01)
            channels = [] if is_vlsd else _collect_channels(b, cg["links"][1])
            groups.append({
                "record_id": record_id, "cycle_count": cycle_count,
                "rec_size": data_bytes + inval_bytes, "data_bytes": data_bytes,
                "channels": channels, "acq": _text(b, cg["links"][2]), "vlsd": is_vlsd,
            })
            cg_off = cg["links"][0]

        data = _read_data_block(b, dg_data)
        total += _decode_dg(b, store, data, groups, rec_id_size)
        dg_off = dg_next

    if total == 0:
        raise ValueError("MDF: no CAN frames found (only CAN bus-logging files are supported).")
    return store


def _classify(b, g):
    def find(suffix):
        for c in g["channels"]:
            if c["name"].lower().split(".")[-1] == suffix:
                return c
        return None

    master = next((c for c in g["channels"] if c["type"] in (2, 3)), None) or find("t")
    idc = find("id")
    data_bytes = next((c for c in g["channels"] if c["name"].lower().split(".")[-1] == "databytes"), None)
    name_has_can = "can" in g["acq"].lower() or any("can_" in c["name"].lower() for c in g["channels"])
    is_error = "errorframe" in g["acq"].lower() or any("errorframe" in c["name"].lower() for c in g["channels"])
    if not master or (not idc and not is_error):
        return None
    if not name_has_can and not idc:
        return None
    return {
        "master": master, "id": idc, "data_bytes": data_bytes,
        "ide": find("ide"), "dlc": find("dlc"), "data_length": find("datalength"),
        "dir": find("dir"), "fd": find("edl") or find("fdf"),
        "brs": find("brs"), "esi": find("esi"), "bus": find("buschannel"),
        "is_error": is_error, "conv": _read_conversion(b, master["cc"]),
    }


def _decode_dg(b, store, data, groups, rec_id_size):
    for g in groups:
        g["can"] = None if g["vlsd"] else _classify(b, g)
    added = 0
    n = len(data)
    if rec_id_size == 0:
        if not groups:
            return 0
        g = groups[0]
        pos = 0
        for _ in range(g["cycle_count"]):
            if pos + g["rec_size"] > n:
                break
            added += _emit(store, data, pos, g)
            pos += g["rec_size"]
        return added
    by_id = {g["record_id"]: g for g in groups}
    pos = 0
    while pos + rec_id_size <= n:
        rid = int.from_bytes(data[pos:pos + rec_id_size], "little")
        pos += rec_id_size
        g = by_id.get(rid)
        if not g or pos + g["rec_size"] > n:
            break
        added += _emit(store, data, pos, g)
        pos += g["rec_size"]
    return added


def _emit(store, data, base, g):
    can = g["can"]
    if not can:
        return 0
    t_raw = _read_field(data, base, can["master"])
    t = can["conv"][0] + can["conv"][1] * t_raw if can["conv"] else t_raw
    if can["is_error"] and not can["id"]:
        ch = _read_field(data, base, can["bus"]) if can["bus"] else 0
        store.add(t, 0, False, ch, ERR, None)
        return 1
    idv = _read_field(data, base, can["id"]) & 0xFFFFFFFF
    if can["ide"]:
        ext = _read_field(data, base, can["ide"]) != 0
    else:
        ext = bool(idv & 0x80000000)
    idv &= 0x1FFFFFFF
    ch = _read_field(data, base, can["bus"]) if can["bus"] else 0
    flags = ERR if can["is_error"] else 0
    if can["fd"] and _read_field(data, base, can["fd"]):
        flags |= FD
    if can["brs"] and _read_field(data, base, can["brs"]):
        flags |= BRS
    if can["esi"] and _read_field(data, base, can["esi"]):
        flags |= ESI
    if can["dir"] and _read_field(data, base, can["dir"]):
        flags |= TX
    fd = bool(flags & FD)
    if can["data_length"]:
        length = _read_field(data, base, can["data_length"])
    elif can["dlc"]:
        length = dlc_to_length(_read_field(data, base, can["dlc"]), fd)
    else:
        length = can["data_bytes"]["bit_count"] // 8 if can["data_bytes"] else 0
    length = max(0, min(64, int(length)))
    payload = None
    if can["data_bytes"]:
        off = base + can["data_bytes"]["byte_offset"]
        width = can["data_bytes"]["bit_count"] // 8
        payload = data[off:off + min(length, width)]
    store.add(t, idv, ext, ch, flags, payload)
    return 1


register_format(Format(
    name="ASAM MDF v4 (MF4)",
    extensions=["mf4", "mdf"],
    sniff=lambda head, name: head[:3] == b"MDF",
    read=parse_mdf,
))
